### ISSUE TITLE
[Preview 8] Blazor localization

### ISSUE THREAD

**Issue by rynowak** [aspnet/AspNetCore.Docs#13436]
# Summary

To localize a server-side Blazor application use the localization middleware (https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-2.2#localization-middleware) to select the appropriate culture for each user. You can you use any scheme that you want to select the culture, see the localization middleware documentation for examples.

## Cookies

As as special consideration for server-size Blazor we recommend setting the culture in a cookie as part of your host page. This will ensure that the websocket connection can propagate the culture correctly. Localization schemes based on the URL path or query string may not be able to work with websockets, so using the cookie as a fallback is useful.

Here's an example of how to set the current culture in a cookie that can be read by the localization middleware.

```C#
    public class HostModel : PageModel
    {
        public void OnGet()
        {
            HttpContext.Response.Cookies.Append(
                    CookieRequestCultureProvider.DefaultCookieName,
                    CookieRequestCultureProvider.MakeCookieValue(
                        new RequestCulture(
                            CultureInfo.CurrentCulture,
                            CultureInfo.CurrentUICulture)));
        }
    }
```

Here's an explanation of how this works:
1. The browser sends an initial HTTP request to the application
2. The culture is assigned by the localization middleware
3. The `_Host.cshtml` page persists the culture in a cookie as part of the response
4. The browser opens a web socket to being an interactive server-side Blazor session
5. The localization middleware reads the cookie and assign the culture
6. The server-side Blazor session begins with the correct culture

## Providing UI to choose the culture

To provide UI to allow users of your application to select the culture, we recommend using a redirect-based approach. This means that a new server-side Blazor session will start, and you should persist any state you want to retain, similar to what happens when a user logs in.

In order to make this work, you will need an HTTP endpoint on the server to change the users stored locale, and perform the redirect. Here's an example you can customize:

```C#
    [Route("[controller]/[action]")]
    public class CultureController : Controller
    {
        public IActionResult SetCulture(string culture, string redirectUri)
        {
            if (culture != null)
            {
                HttpContext.Response.Cookies.Append(
                    CookieRequestCultureProvider.DefaultCookieName,
                    CookieRequestCultureProvider.MakeCookieValue(new RequestCulture(culture)));
            }

            return LocalRedirect(redirectUri);
        }
    }
```

This controller will accept the culture and a redirect URI in the query string, and will update the users' stored culture cookie before redirecting them. Note: It's very important to use the `LocalRedirect` action result here for security reasons. 

Here's a simplified sample of how to perform this redirection flow from a component:

```razor
@inject IUriHelper UriHelper

<h3>Select your language</h3>
<select @onchange="OnSelected" id="culture-selector">
    <option>Select...</option>
    <option value="en-US">English</option>
    <option value="fr-FR">Français</option>
</select>
@code {
    double textNumber;

    void OnSelected(UIChangeEventArgs e)
    {
        var culture = (string)e.Value;
        var uri = new Uri(UriHelper.GetAbsoluteUri()).GetComponents(UriComponents.PathAndQuery, UriFormat.Unescaped);
        var query = $"?culture={Uri.EscapeDataString(culture)}&redirectUri={Uri.EscapeDataString(uri)}";
        UriHelper.NavigateTo("/Culture/SetCulture" + query, forceLoad: true);
    }
}
```

## Using localization features from Blazor

Inside Blazor applications you can use existing localization and globalization features such as .NET's resources system, or culture-specific formatting of numbers and dates. Blazor's `@bind` functionality performs globalization based on the current culture.

A limited set of ASP.NET Core's localization features are supported in Blazor at the time of writing. `IStringLocalizer<>` can be used with Blazor. `IHtmlLocalizer<>`, `IViewLocalizer<>` and Data Annotations localization are ASP.NET Core MVC features and cannot be used with Blazor. https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-2.2#make-the-apps-content-localizable

**Comment by rynowak** [comment]
@guardrex - updated the main post with some info. We might want to borrow some content from https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-2.2#make-the-apps-content-localizable since it has a lot of general info about what localization is and how resources work.

**Comment by guardrex** [comment]
@danroth27 @rynowak ... I'm working on this and the globalization issue now. On the first pass, I'll take your, Ryan, guidance and roll these into the *Components* topic.

The *Components* topic is getting rather *loooong*. It's fine if the content is long&mdash;that's rarely a serious problem. However, the sidebar TOC is eventually going to be a bear to use at some point. I'm not saying that we need to do anything right this second, but let's keep an :eye: on it.

The current *Components* topic sidebar TOC ... 

<img width="138" alt="Capture" src="https://user-images.githubusercontent.com/1622880/62156940-a228c280-b2d1-11e9-9e39-f8e66ac97340.PNG">

### 😅

**Comment by endeffects** [comment]
When we can expect a solution for blazor client side applications?

**Comment by guardrex** [comment]
Note to self: Cross-link 'how to set culture' in globalization remarks when this is worked.

https://github.com/aspnet/AspNetCore.Docs/issues/13435#issue-471467030

**Comment by guardrex** [comment]
Resolved by #13724
